src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/ResourceScopeImpl.java

```diff
@@ -91,72 +89,37 @@ void addInternal(ResourceList.ResourceCleanup resource) {
         }
     }
 
-    protected ResourceScopeImpl(Cleaner cleaner, ResourceList resourceList) {
+    protected ResourceScopeImpl(ResourceList resourceList, Cleaner cleaner) {
         this.resourceList = resourceList;
-        if (cleaner != null) {
-            cleaner.register(this, resourceList);
-        }
-    }
-
-    public static ResourceScopeImpl createImplicitScope() {
-        return new ImplicitScopeImpl(CleanerFactory.cleaner());
+        cleanable = (cleaner != null) ?
+            cleaner.register(this, resourceList) : null;
     }
 
     public static ResourceScopeImpl createConfined(Thread thread, Cleaner cleaner) {
         return new ConfinedScope(thread, cleaner);
     }
 
-    /**
-     * Creates a confined memory scope with given attachment and cleanup action. The returned scope
-     * is assumed to be confined on the current thread.
-     * @return a confined memory scope
-     */
-    public static ResourceScopeImpl createConfined(Cleaner cleaner) {
-        return new ConfinedScope(Thread.currentThread(), cleaner);
-    }
-
-    /**
-     * Creates a shared memory scope with given attachment and cleanup action.
-     * @return a shared memory scope
-     */
     public static ResourceScopeImpl createShared(Cleaner cleaner) {
         return new SharedScope(cleaner);
     }
 
-    private final void release0(HandleImpl handle) {
-        try {
-            Objects.requireNonNull(handle);
-            if (handle.scope() != this) {
-                throw new IllegalArgumentException("Cannot release an handle acquired from another scope");
-            }
-            handle.release();
-        } finally {
-            Reference.reachabilityFence(this);
-        }
-    }
-
     @Override
-    public final void release(ResourceScope.Handle handle) {
-        release0((HandleImpl)handle);
+    public MemorySegment allocate(long bytesSize, long bytesAlignment) {
+        return MemorySegment.allocateNative(bytesSize, bytesAlignment, this);
     }
 
-    @Override
-    public final void release(ScopedMemoryAccess.Scope.Handle handle) {
-        release0((HandleImpl)handle);
-    }
+    public abstract void release0();
 
-    @Override
-    public abstract HandleImpl acquire();
+    public abstract void acquire0();
 
-    /**
-     * Internal interface used to implement resource scope handles.
-     */
-    public non-sealed interface HandleImpl extends ResourceScope.Handle, ScopedMemoryAccess.Scope.Handle {
-
-        @Override
-        ResourceScopeImpl scope();
-
-        void release();
+    @Override
+    public void keepAlive(ResourceScope target) {
+        if (target == this) {
+            throw new IllegalArgumentException("Invalid target scope.");
+        }
+        ResourceScopeImpl targetImpl = (ResourceScopeImpl)target;
+        targetImpl.acquire0();
+        addCloseAction(targetImpl::release0);
```

Comment:
Maybe this should explicitly check if target is `null` (though the call to `acquire0` would also produce an NPE, the stack trace having Objects::requireNonNull in there would make the error more obvious I think).
```suggestion
    public void keepAlive(ResourceScope target) {
        Objects.requireNonNull(target);
        if (target == this) {
            throw new IllegalArgumentException("Invalid target scope.");
        }
        ResourceScopeImpl targetImpl = (ResourceScopeImpl)target;
        targetImpl.acquire0();
        addCloseAction(targetImpl::release0);
```